### ISSUE TITLE
Consider supporting more query operators inside EF.CompileQuery

### ISSUE THREAD

**Issue by divega** [aspnet/EntityFrameworkCore#14547]
For instance, ToList, or ToArray. I am not sure how the EF.CompileQuery/EF.CompileAsyncQuery is going to evolve with [the adoption of the new `IAsyncEnumerable<T>`](https://github.com/aspnet/EntityFrameworkCore/issues/12047), but if it stays similar to what we have today, we could add logic to support what @cwoodruff described in https://twitter.com/cwoodruff/status/1090236542696779778:

``` csharp
private static readonly Func<ChinookContext, Task<List<Album>>> _queryGetAllAlbums =
    EF.CompileAsyncQuery((ChinookContext db) => db.Album.ToList());
```

This is matching an overload of CompileAsyncQuery that returns Task<TResult>. These overloads are designed to work for methods that return a single element, e.g. a query that ends in First(). We didn't expect customers to use them to get a List<T>. 

AFAIK, if you want to get a list asynchronously from a compiled query, you have to create the compiled query calling an overload of CompileAsyncQuery that takes a lambda that returns an IQueryable<T>, and then call ToListAsync() each time on the result of invoking the compiled query delegate, e.g.:

``` csharp
private static readonly Func<ChinookContext, Task<List<Album>>> _queryGetAllAlbums =
    EF.CompileAsyncQuery((ChinookContext db) => db.Album);
...
var results = await _queryGetAllAlbums(db).ToListAsync();
```

**Comment by cwoodruff** [comment]
Thanks @divega for the help!

**Comment by cwoodruff** [comment]
@divega I had to use the following for the compiled query:

private static readonly Func<ChinookContext, AsyncEnumerable<Album>> _queryGetAllAlbums =
            EF.CompileAsyncQuery((ChinookContext db) => db.Album);

**Comment by divega** [comment]
@cwoodruff correct. I forgot to update the signature of the method. Anyway, I have created 
https://github.com/aspnet/EntityFrameworkCore/issues/14551 to possibly revisit/simplify how the API works with the type system.

**Comment by msmolka** [comment]
@divega @cwoodruff Do you know how to use it in EF Core 3. There is no `AsyncEnumerable` there.
